fsck/fsck.c

```diff
@@ -670,58 +670,72 @@ static int read_boot_region(struct exfat_blk_dev *bd, struct pbr **pbr,
 		return -ENOMEM;
 	}
 
+	if (main_sect)
+		bs_offset = BOOT_SEC_IDX;
+	else
+		bs_offset = BACKUP_BOOT_SEC_IDX;
+
 	if (exfat_read(bd->dev_fd, bs, sizeof(*bs),
-			bs_offset * bd->sector_size) != (ssize_t)sizeof(*bs)) {
+			bs_offset * sect_size) != (ssize_t)sizeof(*bs)) {
 		exfat_err("failed to read a boot sector\n");
 		ret = -EIO;
 		goto err;
 	}
 
 	if (memcmp(bs->bpb.oem_name, "EXFAT   ", 8) != 0) {
-		exfat_err("failed to find exfat file system.\n");
+		if (main_sect)
+			exfat_err("failed to find exfat file system\n");
 		goto err;
 	}
 
-	ret = boot_region_checksum(bd, bs_offset);
+	ret = boot_region_checksum(bd, bs_offset, sect_size);
 	if (ret < 0)
 		goto err;
 
 	ret = -EINVAL;
 	if (EXFAT_SECTOR_SIZE(bs) < 512 || EXFAT_SECTOR_SIZE(bs) > 4 * KB) {
-		exfat_err("too small or big sector size: %d\n",
-				EXFAT_SECTOR_SIZE(bs));
+		if (main_sect)
+			exfat_err("too small or big sector size: %d\n",
+				  EXFAT_SECTOR_SIZE(bs));
 		goto err;
 	}
 
 	if (EXFAT_CLUSTER_SIZE(bs) > 32 * MB) {
-		exfat_err("too big cluster size: %d\n", EXFAT_CLUSTER_SIZE(bs));
+		if (main_sect)
+			exfat_err("too big cluster size: %d\n",
+				  EXFAT_CLUSTER_SIZE(bs));
 		goto err;
 	}
 
 	if (bs->bsx.fs_version[1] != 1 || bs->bsx.fs_version[0] != 0) {
-		exfat_err("unsupported exfat version: %d.%d\n",
-				bs->bsx.fs_version[1], bs->bsx.fs_version[0]);
+		if (main_sect)
+			exfat_err("unsupported exfat version: %d.%d\n",
+				  bs->bsx.fs_version[1], bs->bsx.fs_version[0]);
 		goto err;
 	}
 
 	if (bs->bsx.num_fats != 1) {
-		exfat_err("unsupported FAT count: %d\n", bs->bsx.num_fats);
+		if (main_sect)
+			exfat_err("unsupported FAT count: %d\n",
+				  bs->bsx.num_fats);
 		goto err;
 	}
 
 	if (le64_to_cpu(bs->bsx.vol_length) * EXFAT_SECTOR_SIZE(bs) >
 			bd->size) {
-		exfat_err("too large sector count: %" PRIu64 ", expected: %llu\n",
-				le64_to_cpu(bs->bsx.vol_length),
-				bd->num_sectors);
+		if (main_sect)
+			exfat_err("too large sector count: %" PRIu64 ", expected: %llu\n",
+				  le64_to_cpu(bs->bsx.vol_length),
+				  bd->num_sectors);
 		goto err;
 	}
 
 	if (le32_to_cpu(bs->bsx.clu_count) * EXFAT_CLUSTER_SIZE(bs) >
 			bd->size) {
-		exfat_err("too large cluster count: %u, expected: %u\n",
-				le32_to_cpu(bs->bsx.clu_count),
-				bd->num_clusters);
+		if (main_sect)
```

Comment:
why does it print error message only if main_sect is true ?

Reply:
Because this function is called many times to read and verify the backup boot region with different sector sizes for restoring the main boot region, these error messages is too verbose for this restoration.

Reply:
I think that we should add bool silent variable as an argument of this function.

Reply:
I will declare a local boolean variable for that.

Reply:
I don't know how you do that.